app/api/routers/token.py

```diff
@@ -177,25 +182,155 @@ def get_token_holders(
             IDXLockedPosition.account_address,
         )
     )
+    total = session.scalar(select(func.count()).select_from(stmt.subquery()))
+
     if request_query.exclude_owner is True:
         stmt = stmt.where(IDXPosition.account_address != listed_token.owner_address)
+    count = session.scalar(select(func.count()).select_from(stmt.subquery()))
+
+    # Pagination
+    if limit is not None:
+        stmt = stmt.limit(limit)
+    if offset is not None:
+        stmt = stmt.offset(offset)
 
     holders: Sequence[tuple[IDXPosition, int | None]] = session.execute(
         stmt.order_by(desc(IDXPosition.created))
     ).all()
 
-    resp_body = [
-        {
-            "token_address": holder[0].token_address,
-            "account_address": holder[0].account_address,
-            "amount": holder[0].balance,
-            "pending_transfer": holder[0].pending_transfer,
-            "exchange_balance": holder[0].exchange_balance,
-            "exchange_commitment": holder[0].exchange_commitment,
-            "locked": holder[1] if holder[1] else 0,
-        }
-        for holder in holders
-    ]
+    resp_body = {
+        "result_set": {
+            "count": count,
+            "offset": offset,
+            "limit": limit,
+            "total": total,
+        },
+        "token_holder_list": [
+            {
+                "token_address": holder[0].token_address,
+                "account_address": holder[0].account_address,
+                "amount": holder[0].balance if holder[0].balance else 0,
+                "pending_transfer": holder[0].pending_transfer
+                if holder[0].pending_transfer
+                else 0,
+                "exchange_balance": holder[0].exchange_balance
+                if holder[0].exchange_balance
+                else 0,
+                "exchange_commitment": holder[0].exchange_commitment
+                if holder[0].exchange_commitment
+                else 0,
+                "locked": holder[1] if holder[1] else 0,
+            }
+            for holder in holders
+        ],
+    }
+
+    return json_response({**SuccessResponse.default(), "data": resp_body})
+
+
+@router.post(
+    "/{token_address}/Holders/Search",
+    summary="Search Token holders",
+    operation_id="SearchTokenHolders",
+    response_model=GenericSuccessResponse[TokenHoldersResponse],
+    responses=get_routers_responses(DataNotExistsError, InvalidParameterError),
+)
+def search_token_holders(
+    session: DBSession,
+    data: SearchTokenHoldersRequest,
+    token_address: Annotated[
+        ValidatedEthereumAddress, Path(description="Token address")
+    ],
+):
+    """
+    Endpoint: /Token/{contract_address}/Holders/Search
+    """
```

Comment:
Existing APIs have similar issues, but I think it would be better to write an explanation of the API as a comment.